### ISSUE TITLE
Upload of large files is too slow

### ISSUE THREAD

**Issue by jmlero** [Azure/azure-sdk-for-python#294]
Upload of large files is slow. Upload a file of 10GB takes more than 30 minutes while upload the same file from the same machine using the tool AzCopy takes around 3 minutes.

Maybe this can be solved with the implementation of this feature: 
https://github.com/Azure/azure-sdk-for-python/issues/237
Or with the possibility of select the size of the chunk (is that possible right now?).

Any other ideas for improve the upload of large files?

Thanks in advance,
Best regards,

**Comment by huguesv** [comment]
In a prototype of parallel uploads, ie. #237, we've seen significant improvements.  This is our current plan to improve perf.  This is now very near the top of our TODO list.  We should have something people can try in the coming weeks.
Last time I checked, we were using the max chunk size allowed by Azure.

**Comment by jmlero** [comment]
Thanks a lot for your effort and the development of this SDK. We want to try the next release with the new improvements.

Regards,

**Comment by huguesv** [comment]
I just merged my changes to master. The high level functions to download and upload blobs now have a max_connections parameter that you can set.  For now it defaults to 1 connection, haven't decided on what we should make the default.

I haven't done a lot of speed tests, but it's a definite improvement (I get 50MB/sec upload, 70MB/sec download from an Azure VM in the same data center as the storage account).  It also has built in retries, so that a single block failure won't immediately fail the whole upload.

If you do try it out, please let us know!

**Comment by jmlero** [comment]
I did some test, here are my result:

I tried to upload two single files of 10GB and 100GB using AzCopy, the framework xplat-cli and this framework using 8 and 16 max_connections:

Using AzCopy:
10GB: 3m
100GB: 34m

Using xpat-cli:
10GB: 9m
100GB 89m

Using the previous python sdk:
10GB: 38
100GB: 376

Using the last python sdk:
8 max_connections:
10GB: 4m
100GB: 45m

16 max_connections:
10GB: 3m
100GB: 25m

-All the times are expressed in minutes.
-All the test were done using the same machine using Scientific Linux 6.5 (except for AzCopy using Windows 7) and python 2.7.
-The times are more or less the same for downloading data.

In conclusion, in our test experiments using large single files, the python sdk is faster than AzCopy using more than 8 connections.

Good job, thanks a lot.

Best regards,

**Comment by huguesv** [comment]
This is great to hear, thanks for sharing your results!

**Comment by jehine-MSFT** [comment]
Thanks @jmlero for the feedback! Sounds like we are making great progress for the python storage library. We hope that by the time we publish a GA version we can improve these numbers even more.

Thanks again!
